packages/caver-wallet/src/keyringContainer.js

```diff
@@ -0,0 +1,308 @@
+        }
+        return this
+    }
+
+    /**
+     * creates a keyring instance with given parameters and adds it to the keyringContainer.
+     *
+     * @param {string} address The address of the keyring.
+     * @param {string|Array.<string>|Array.<Array.<string>>} key Private key string(s) to use in keyring. If different keys are used for each role, key must be defined as a two-dimensional array.
+     * @return {Keyring}
+     */
+    newKeyring(address, key) {
+        // The format of key parameter can be
+        // 1. single private key string   => `0x{private key}`
+        // 2. multiple private key string =>[`0x{private key}`, `0x{private key}`, ...]
+        // 3. role based private keys     => [[`0x{private key}`, `0x{private key}`, ...], [], [`0x{private key}`]]
+
+        let keyring
+
+        if (_.isString(key)) keyring = Keyring.createWithSingleKey(address, key)
+
+        if (_.isArray(key)) {
+            if (key.length === 0) throw new Error(`Insufficient private key information: Empty array`)
+            if (_.isArray(key[0])) {
+                keyring = Keyring.createWithRoleBasedKey(address, key)
+            } else {
+                keyring = Keyring.createWithMultipleKey(address, key)
+            }
+        }
+
+        if (!(keyring instanceof Keyring)) throw new Error(`Unsupported type : ${typeof key}`)
+
+        return this.add(keyring)
+    }
+
+    /**
+     * updates the keyring inside the keyringContainer.
+     * Query the keyring to be updated from keyringContainer with the keyring's address,
+     * and an error occurs when the keyring is not found in the keyringContainer.
+     *
+     * @param {Keyring} keyring The keyring with new key.
+     * @return {Keyring}
+     */
+    updateKeyring(keyring) {
+        const founded = this._addressKeyringMap.get(keyring.address.toLowerCase())
+        if (founded === undefined) throw new Error(`Failed to find keyring to update`)
+
+        founded.key = keyring.copy().key
+        return founded
+    }
+
+    /**
+     * Get the keyring in container corresponding to the address
+     *
+     * @param {string} address The address of keyring to query.
+     * @return {Keyring}
+     */
+    getKeyring(address) {
+        if (!utils.isAddress(address))
+            throw new Error(
+                `Invalid address ${address}. To get keyring from wallet, you need to pass a valid address string as a parameter.`
+            )
+
+        const founded = this._addressKeyringMap.get(address.toLowerCase())
+
+        return founded
+    }
+
+    /**
+     * adds a keyring to the keyringContainer.
+     *
+     * @param {Keyring} keyring A keyring instance to add to keyringContainer.
+     * @return {Keyring}
+     */
+    add(keyring) {
+        if (this._addressKeyringMap.get(keyring.address.toLowerCase()) !== undefined)
+            throw new Error(`Duplicate Account ${keyring.address}`)
+
+        const keyringToAdd = keyring.copy()
+
+        this._length++
```

Comment:
Do we have enough test code to check this value? this length might be inconsistent with the length of the map for some cases.

Reply:
I think keyringContainer doesn't need to manage length separately, so i removed '_length' in KeyringContainer.